physlr/physlr.py

```diff
@@ -1021,107 +1020,118 @@ def assign_read_molecule(mxs, moltomxs, mol_counts, bx):
-            return u, {v: i for i, vs in enumerate(components) if len(vs) > 1 for v in vs}
-        partition = louvain.best_partition(sub_graph)
-        if not partition:
-            return u, {}
-        multinode_partitions_set = {}
-        for com in set(partition.values()):
-            list_nodes = [nodes for nodes in partition.keys() if partition[nodes] == com]
-            if len(list_nodes) > 1:
-                multinode_partition = {v: com for v in list_nodes}
-                multinode_partitions_set.update(multinode_partition)
-        return u, multinode_partitions_set
+        if len(node_set) > 1:
+            if not init_communities:
+                partition = louvain.best_partition(g.subgraph(node_set))
+            else:
+                partition = louvain.best_partition(g.subgraph(node_set), init_communities)
+            return [{nodes for nodes in partition.keys() if partition[nodes] == com}
+                    for com in set(partition.values())]
+        return []
 
     @staticmethod
-    def determine_molecules_cosine_of_squared(g, u):
-        "Square the adjacency matrix and then use cosine similarity to detect communities."
+    def detect_communities_cosine_of_squared(g, node_set):
+        """
+        Square the adjacency matrix and then use cosine similarity to detect communities.
+        Return communities.
+        """
         import scipy as sp
         import numpy as np
         from sklearn.metrics.pairwise import cosine_similarity
 
-        cut_vertices = set(nx.articulation_points(g.subgraph(g.neighbors(u))))
-        components = list(nx.connected_components(g.subgraph(set(g.neighbors(u)) - cut_vertices)))
-        components.sort(key=len, reverse=True)
         communities = []
-        for comp in components:
-            if len(comp) > 1:
-                sub_graph = g.subgraph(comp)
-                adj_array = nx.adjacency_matrix(sub_graph).toarray()
-                new_adj = np.multiply(
-                    cosine_similarity(
-                        sp.linalg.blas.sgemm(1.0, adj_array, adj_array)) >= 0.75, adj_array)
-                edges_to_remove = np.argwhere(new_adj != adj_array)
-                barcode_dict = dict(zip(range(len(sub_graph)), list(sub_graph.nodes)))
-                edges_to_remove_barcode = [(barcode_dict[i], barcode_dict[j])
-                                           for i, j in edges_to_remove]
-                sub_graph_copy = nx.Graph(sub_graph)
-                sub_graph_copy.remove_edges_from(edges_to_remove_barcode)
-                cos_components = list(nx.connected_components(sub_graph_copy))
-                cos_components.sort(key=len, reverse=True)
-                for com in cos_components:
-                    if len(com) > 1:
-                        communities.append(com)
-        return u, {v: i for i, vs in enumerate(communities) if len(vs) > 1 for v in vs}
+        if len(node_set) > 1:
```

Comment:
Remove special cases unless truly needed. Use early return if it is truly needed.
```python
if len(node_set) < 2:
  return []
```